### ISSUE TITLE
"No TileSource was able to open" with some IIIF sources but not others

### ISSUE THREAD

**Issue by craigberry** [openseadragon/openseadragon#2325]
The attached standalone reproducer demonstrates a problem I've encountered with a particular set of IIIF images.  Same problem with OpenSeadragon 3.1.0 and 4.0.0.  The code I'm using here works fine for many thousands of images from a number of different sources.  But for a particular set of images from the University of Illinois at Urbana-Champaign, it throws an error that looks like:

![image](https://user-images.githubusercontent.com/93371/227781862-7fb722d2-86c5-4588-997e-cc846628a98c.png)

If I go to line 978 in tilesource.js, it's the determineType function that does not find a matching TileSource type.  Setting a breakpoint in that function reveals that the data parameter has not been populated with the contents of the info.json file.  So it appears the info.json file is not getting correctly processed, but I don't see anything obviously wrong with that file.

So I'm lost about whether I'm doing something wrong (but then why does it work in most cases), if there is some dodgy data in the UofI info.json files (but they work fine with their viewer), or if there is a bug in OpenSeadragon.  All seem unlikely.  Any insight appreciated.

[uofi_image_problem.html.zip](https://github.com/openseadragon/openseadragon/files/11071714/uofi_image_problem.html.zip)

**Comment by pearcetm** [comment]
> The fix may be as simple as changing the XML regex from `/\s*<.*/` to `/^\s*<.*/`. As far as I know, well-formed XML cannot have non-whitespace text nodes outside of the root node, so there should always be an element (or xml declaration or processing instruction) at the beginning.

This sounds reasonable to me - as far as I am aware, this should be true, and the proposed change to the `regex` shouldn't break anything that was previously functional.



> And I think parseXml will throw an error if the data is not well-formed XML.

I would have thought the same, but I think this is not true. It seems that non-microsoft browsers (since at least 2015) instead generate an XML document that details the error (but does not throw), rather than throwing an error.

> I assume the regex is because `Content-Type` headers are not always present? Would it be safe to believe them if they _are_ present, and then fall back to the regexen if there's no `application/xml` or `application/json`?

I don't know the history or the logic behind the current implementation, so I can't really predict whether defaulting to `Content-Type` headers would cause unintended side effects. I'd hope not, but...?

**Comment by craigberry** [comment]
> > And I think parseXml will throw an error if the data is not well-formed XML.
> 
> I would have thought the same, but I think this is not true. It seems that non-microsoft browsers (since at least 2015) instead generate an XML document that details the error (but does not throw), rather than throwing an error.
> 

The following demonstrates that `parseXML` throws an error on malformed XML input with current jQuery.  I tried it on recent Safari, Firefox, and Brave (chromium-based) browsers.  While that is not the whole universe, it's a pretty good sampling.  Are you still trying to support IE?  I can give that a try later.  The [parseXML docs](https://api.jquery.com/jQuery.parseXML/) say the input is "a well-formed XML string to be parsed"; they don't specify how malformed input is handled.

```html
<!DOCTYPE html>
<html>
<head>
    <title>Test jQuery parseXml function</title>
    <script type="text/javascript" src='https://code.jquery.com/jquery-3.6.4.js'></script>
</head>
<body>
<div>
    <p id="goodxml"></p>
    <p id="badxml"></p>
</div>
</div>
<script type="text/javascript">
    var goodxml = "<tag>Something well-formed this way comes.</tag>";
    var badxml = "  garbage <tag>Something malformed this way comes.</tag>";
    var data;

    try {
        data = jQuery.parseXML( goodxml );
    }
    catch(e) {
        console.log('goodxml', e);
        data = "Thought the XML was good, but it wasn't."
    }
    $xml = $(data);
    $text = $xml.find('tag');
    $('#goodxml').append( $text.text() );

    try {
        data = jQuery.parseXML( badxml );
    }
    catch(e) {
        console.log('badxml', e);
        data =  badxml + ": " + e.message;
    }
    $('#badxml').append( data );

</script>
</body>
</html>
```

**Comment by pearcetm** [comment]
> The following demonstrates that `parseXML` throws an error on malformed XML input with current jQuery.

Looks like the version used by OpenSeadragon is older, because calling `OpenSeadragon.parseXml( badxml )` generates the following, with no error thrown:

```
<html xmlns="http://www.w3.org/1999/xhtml">
    <body>
        <parsererror style="display: block; white-space: pre; border: 2px solid #c77; padding: 0 1em 0 1em; margin: 1em; background-color: #fdd; color: black">
             <h3>This page contains the following errors:</h3>
             <div style="font-family:monospace;font-size:12px">error on line 1 at column 3: Document is empty</div>
             <h3>Below is a rendering of the page up to the first error.</h3>
        </parsererror>
    </body>
</html>
```

It probably wouldn't be a bad idea to update the `parseXml` function (though by itself it wouldn't solve your problem).


> Are you still trying to support IE?

IE support is being phased out, but there's one more release to go before that point: https://github.com/openseadragon/openseadragon/issues/2302

**Comment by iangilman** [comment]
Fixed by #2328.
